staging/src/k8s.io/apiserver/pkg/endpoints/openapi/cacher.go

```diff
@@ -0,0 +1,268 @@
+	if err != nil {
+		return err
+	}
+
+	// The schema hasn't changed so there's nothing to update
+	if httpStatus == http.StatusNotModified {
+		return nil
+	}
+
+	s.rwMutex.Lock()
+	defer s.rwMutex.Unlock()
+
+	// Check etag again after taking the write lock in case lastEtag
+	// was updated while waiting for the write lock
+	if newEtag == s.lastEtag {
+		return nil
+	}
+
+	// Convert raw spec to our Resource type
+	schema, err := s.parser.Parse(schemaBytes)
+	if err != nil {
+		return err
+	}
+
+	s.openAPISchema = schema
+	s.lastEtag = newEtag
+	return nil
+}
+
+func (s *SchemaCacher) getLastEtag() string {
+	s.rwMutex.RLock()
+	defer s.rwMutex.RUnlock()
+	return s.lastEtag
+}
+
+// Downloader sepcifies where to download an OpenAPI schema from
+type Downloader interface {
+	// DownloadSchema attempts to get the OpenAPI schema if it has changed.
+	DownloadSchema(lastEtag string) (
+		schemaBytes []byte,
+		newEtag string,
+		httpStatus int,
+		err error)
+}
+
+type handlerDownloader struct {
+	handler http.Handler
+	path    string
+}
+
+var _ Downloader = &handlerDownloader{}
+
+// DownloadSchema implements Downloader
+func (d *handlerDownloader) DownloadSchema(lastEtag string) (schemaBytes []byte, newEtag string, httpStatus int, err error) {
+	req, err := http.NewRequest("GET", d.path, nil)
+	if err != nil {
+		return nil, "", 0, err
+	}
+	req.Header.Add("Accept", "application/com.github.proto-openapi.spec.v2@v1.0+protobuf")
+	if len(lastEtag) > 0 {
+		req.Header.Add("If-None-Match", lastEtag)
+	}
+
+	writer := utilresponsewriter.NewInMemoryResponseWriter()
+	d.handler.ServeHTTP(writer, req)
+
+	switch writer.RespCode {
+	case http.StatusNotModified:
+		if len(lastEtag) == 0 {
+			return nil, lastEtag, http.StatusNotModified, fmt.Errorf("http.StatusNotModified is not allowed in absence of etag")
+		}
+		return nil, lastEtag, http.StatusNotModified, nil
+	case http.StatusOK:
+		schemaBytes = writer.Data
+		newEtag = writer.Header().Get("Etag")
+		return schemaBytes, newEtag, http.StatusOK, nil
+	default:
+		return nil, "", 0, fmt.Errorf("failed to retrieve openAPI spec, http error: %s", writer.String())
+	}
+}
+
+// This interface is to here to make unit testing easier
+type Parser interface {
+	// Parse converts a byte array to a Resouces object for easier schema lookups
+	Parse(doc []byte) (Resources, error)
+}
+
+type schemaParser struct{}
+
+var _ Parser = &schemaParser{}
+
+// Parse implements Parser
+func (p *schemaParser) Parse(raw []byte) (Resources, error) {
+	doc := &openapi_v2.Document{}
+	err := golangproto.Unmarshal(raw, doc)
+	if err != nil {
+		return nil, err
+	}
+	models, err := proto.NewOpenAPIData(doc)
+	if err != nil {
+		return nil, err
+	}
+
+	resources := map[schema.GroupVersionKind]string{}
+	for _, modelName := range models.ListModels() {
+		model := models.LookupModel(modelName)
+		if model == nil {
+			panic("ListModels returns a model that can't be looked-up.")
```

Comment:
Return an error instead?

Reply:
Hmm this one I just copied from kubectl, but I can definitely change it

Reply:
Yeah, there is something going seriously wrong with the code if that happens. I think panics are more appropriate for these use-cases.

Reply:
This means I could potentially make apiserver crashloop by aggregating a server which serves a bad spec.

Reply:
it's not a disaster if kubectl crashes on bad input but it *is* a disaster if apiserver does.

Reply:
Then you should probably just return the condition altogether. That's equivalent to writing:

```
for key := range somemap {
    if _, ok := somemap[key]; !ok {
       panic(...)
    }
}
```
Nobody checks that condition for obvious reasons, we shouldn't return an error.